### PR TITLE
Cache the list of nearby intersections to wake up for uber-turns. #497

### PR DESCRIPTION
**Pull request by dabreegster**
This shouldn't have any behavioral change.

@BruceBrown

### CODE REVIEW

File: sim/src/mechanics/intersection.rs [hunk 1]
```diff
@@ -561,6 +570,14 @@ impl IntersectionSimState {
                 }
                 (None, None) => {}
             }
+
+            // It's unlikely, but the player might create/destroy traffic signals close together and
+            // change the uber-turns that exist. To be safe, recalculate everywhere.
+            state.uber_turn_neighbors.clear();
```
Comment (BruceBrown) [inline sim/src/mechanics/intersection.rs:576]: It actually will happen a lot. Covert stop to ts and convert midblock crosswalk to degenerate w/ crosswalk.

**Review by BruceBrown — APPROVED** [review]

File: sim/src/mechanics/intersection.rs [hunk 2]
```diff
@@ -561,6 +570,14 @@ impl IntersectionSimState {
...
+            if let Some(mut set) = map_model::IntersectionCluster::autodetect(state.id, map) {
```
Comment (michaelkirk) [inline sim/src/mechanics/intersection.rs:577]: I'm not very familiar with this code... but it seems like `state` is per intersection, right?

So if we edit *this* intersection, we invalidate `self.state.uber_turn_neighbors`. That makes sense to me.

But if editing this intersection changed the uber turns which exist, couldn't this intersection have been newly added to/removed from nearby uber turns - in which case the *other* intersections in that affected uber turn need to invalidate *their* `uber_turn_neighbors` as well? 

Or am I missing something?

Comment (dabreegster) [reply sim/src/mechanics/intersection.rs:577]: > in which case the other intersections in that affected uber turn need to invalidate their uber_turn_neighbors as well

Correct, we potentially need to recalculate `uber_turn_neighbors` for all intersections. That's what this loop does. We iterate over all `for state in self.state.values_mut()` and then totally recalc from scratch.

The calculation does `map_model::IntersectionCluster::autodetect`, which doesn't know about `uber_turn_neighbors` at all; it's not like we have to propagate the change out in a certain order or anything. Detecting clusters of intersections is only based on the map, not any sim state

Comment (michaelkirk) [reply sim/src/mechanics/intersection.rs:577]: Ah, thank you for resolving my confusion. `self` here  is the "per map" `intersection::IntersectionSimState`. `self.state` is **every** per-intersection `state`.

**Pull request merged**